### PR TITLE
Add log parsing to test framework

### PR DESCRIPTION
**Pull request by lwr20**

### CODE REVIEW

File: calico_test/tests/st/utils/docker_host.py [hunk 1]
```diff
@@ -140,6 +142,38 @@ def execute(self, command):
...
+        logger.debug("Final command: %s", command)
+        proc = Popen(command, stdout=PIPE, shell=True)
+
+        try:
+            # Read and return one line at a time until no more data is
+            # returned.
+            for line in proc.stdout:
+                yield line
+        finally:
+            status = proc.wait()
```
Comment (fasaxc) [inline calico_test/tests/st/utils/docker_host.py]: Can you put a timeout on here?  If a process blocks here it's a pain to spot/debug.

File: calico_test/tests/st/utils/log_analyzer.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    "(?P<timestamp>\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}).\d{0,3} "
+    "\\[(?P<loglevel>\w+)\\]"
+    "\\[(?P<pid>\d+)(/\d+)?\\] "
+    "(?P<logtext>.*)"
+)
+
+TIMESTAMP_FORMAT = "%Y-%m-%d %H:%M:%S"
+
+# This is the list of logs we should ignore for all tests.
+# Currently empty, but I'm sure we'll find some to add to it shortly.
```
Comment (fasaxc) [inline calico_test/tests/st/utils/log_analyzer.py]: Out-of-date comment

File: calico_test/tests/st/utils/log_analyzer.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    def append(self, logtext):
+        """
+        Append the text to the end of the current text with a newline
+        separator.
+
+        :param logtext:  The text to append to the log.
+        """
+        self.msg += "\n" + logtext
+
+    def __str__(self):
```
Comment (fasaxc) [inline calico_test/tests/st/utils/log_analyzer.py:64]: Please can we print out the log as it was written?  Breaking it up and reassembling it makes it harder to grep for (and sometimes loses information).

It's also be great to have context either side of the offending log if available.  I've certainly found that was lacking in the OpenStack tests.

File: calico_test/tests/ut/test_log_parsing.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                   str(recursive).lower(), " ".join(options))
+    else:
+        command = "curl -sL http://%s:2379/v2/keys/%s?recursive=%s %s" % \
+                  (ip, path, str(recursive).lower(), " ".join(options))
+    _log.debug("Running: %s", command)
+    rc = subprocess.check_output(command, shell=True)
+    return json.loads(rc.strip())
+
+
+class MultiHostIpfix(TestBase):
```
Comment (fasaxc) [inline calico_test/tests/ut/test_log_parsing.py]: `MultiHostIpfix`?

File: calico_test/tests/st/utils/log_analyzer.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
+from datetime import datetime
```
Comment (fasaxc) [inline calico_test/tests/st/utils/log_analyzer.py:15]: Copyright

File: calico_test/tests/ut/test_log_parsing.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+    @classmethod
+    def tearDownClass(cls):
+        # Tidy up
+        for host in cls.hosts:
+            host.cleanup()
+            del host
+
+    def setUp(self):
+        # This method is REQUIRED or else the TestBase setUp() method will wipe etcd before
```
Comment (fasaxc) [inline calico_test/tests/ut/test_log_parsing.py]: So why inherit from TestBase, does it buy you anything for this test?

File: calico_test/tests/st/utils/log_analyzer.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+import logging
+import re
+
+from tests.st.utils.exceptions import CommandExecError
+
+_log = logging.getLogger(__name__)
+
+FELIX_LOG_FORMAT = (
```
Comment (fasaxc) [inline calico_test/tests/st/utils/log_analyzer.py:23]: Do we only slurp the felix log?

File: calico_test/tests/ut/test_log_parsing.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    for i in range(num_hosts):
+        hostnames.append("host%s" % i)
+    pool = Pool(num_hosts)
+    hosts = pool.map(makehost, hostnames)
+    pool.close()
+    pool.join()
+    return hosts
+
+
+def wipe_etcd():
```
Comment (fasaxc) [inline calico_test/tests/ut/test_log_parsing.py]: Do you use these utility functions?

**Review by fasaxc — CHANGES_REQUESTED** [review]

File: calico_test/tests/st/utils/log_analyzer.py (hunk 7 above)
Comment (lwr20) [reply calico_test/tests/st/utils/log_analyzer.py:23]: At present, yes.  I guess we could extend the logger to also slurp other logs, but felix is a good start, and on a par with other test frameworks.

File: calico_test/tests/st/utils/log_analyzer.py (hunk 5 above)
Comment (lwr20) [reply calico_test/tests/st/utils/log_analyzer.py:15]: Done

File: calico_test/tests/st/utils/log_analyzer.py (hunk 3 above)
Comment (lwr20) [reply calico_test/tests/st/utils/log_analyzer.py:64]: Done

**Review by fasaxc — APPROVED** [review]

**Pull request merged**